lsコマンド1を作成

o-musubi opened the pull request:
lsコマンド、ひとまず必須要件のみ満たしたもの作成しました。
確認と修正お願いします。
(4日前のpush履歴は無視してください、すみません)

o-musubi committed:
ls.rbを作成

o-musubi committed:
lsコマンド1を作成

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,37 @@
+    if displayed_columns[0].nil?
+      puts ' '
+    elsif file[:splited_columns[file[:specific_number]]].nil? && index == file[:specific_number]
+      printf('%-20s', displayed_columns[0])
```
20文字以上のファイル名のときにくっついてしまいます。ファイル名が長くても対応できるようにしてみてください〜

o-musubi replied:
[このページ](https://bootcamp.fjord.jp/questions/707)を読み直し、
毎回最大ファイルから幅を求める必要があるのかな…？と思ったので、
そのように作り変えました。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,37 @@
+    file: Dir.glob('*'),
+    maximum_columns: 3
+  }
+end
```
設定ファイルのような形で定義した、という意図ですかね〜。
確かにそういうこともあるかもしれないですが、実際に設定としてユーザーがカスタマイズ可能としない限り、雑多なものを押し込めただけに見えるので、あまり設計としてはよくないかなと思います。
今後オプションによって、対象ファイルリストが変わったりしますが、これだと対応も難しいかなと思います。
意味のあるまとまりとしてメソッドを設計するように心がけてみてください〜。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,37 @@
+  }
+end
+
+def format_file
```
`format_file` という名前なのにcolumnの数の計算などをしているのはちょっとわかりづらいですね。
通常のアプリケーションやライブラリのコードは書かれる時間より読まれる時間の方が長いです。読む人がわかりやすいような名前を考えてみてください。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,37 @@
+      puts ' '
+    else
+      printf('%-20s', displayed_columns[0])
+    end
```
` printf('%-20s', displayed_columns[0])` が必要なときと `puts ' '` が必要なときでわけたほうがスッキリするのではと思いました。

```ruby
printf('%-20s', displayed_columns[0]) if (printf が必要な条件)
puts ' ' if (puts が必要な条件)
```

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,37 @@
+def main
+  file = format_file
+  file[:splited_columns].each.with_index(1).cycle(file[:column_difference]) do |column, index|
+    displayed_columns = [] << column.shift
```
以下では結局 `displayed_columns[0]` しか使ってなさそうに見えますが、配列にする必要はありそうですか？

o-musubi replied:
なかったです、すみません！
[こちらの参考リンク](https://bootcamp.fjord.jp/pages/148)を見てキューだと思い込んで、配列に入れなければいけないと信じて疑っていませんでした…。
変数に入れる形で修正しました。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,37 @@
+def set_conditions
+  {
+    file: Dir.glob('*'),
+    maximum_columns: 3
```
こういう変わらない値は定数とするとよいです。
https://docs.ruby-lang.org/ja/latest/doc/spec=2fvariables.html#const

- コードの見た目上変わらないことがわかりやすい
- 変更しようとすると警告がでる

o-musubi replied:
定数について、クラス定義がないのでトップレベルに書いています。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,37 @@
+    displayed_columns = [] << column.shift
+    if displayed_columns[0].nil?
+      puts ' '
+    elsif file[:splited_columns[file[:specific_number]]].nil? && index == file[:specific_number]
```
```ruby
file[:splited_columns[file[:specific_number]]].nil?
``` 

この条件、ちょっとおかしいかなと思うんですが、意図した通りに働いていますか？
`file[]` で指定しているキーが `:splited_columns[file[:specific_number]]` になっていますよね。
`file[:specific_number]` は `condition[:splited_columns].size` なので、each_sliceした配列の長さで、ls対象のファイルの数に応じてかわりますが、たとえば3として、  `:splited_columns[3]` がどうなるかというと

https://docs.ruby-lang.org/ja/latest/method/Symbol/i/=5b=5d.html

を見ると「nth 番目の文字を返します。」と書いてあるのでこの場合 `i` になります。ということで、全体としては `file['i']` になるんですが、たぶん間違ってますよね。

o-musubi replied:
解説頂いたとおり、`i`を返す結果は意図したものではありません。
ここは読みなおしたら自分でもなぜこうしたのか思い出せず…すみません。まさにコードをこねくり回して混乱していた部分です。
構造をごっそり見直しました。

yoshitsugu reviewed (commented):
コメントしました！ちょっと混乱していそうなコードに見えるので、あらためて日本語でやりたいことを書き出してもいいかもしれません。

```
1. ファイル名を取得する
2. 3列ごとにならべて表示する
    2-1. ファイル名を3つのグループ(列に対応する)に分割する
    2-2. ...
```

みたいな感じです。

o-musubi committed:
レビューを受けて大幅に構造を修正

o-musubi commented:
レビューいただいた点、自分なりに修正しました。
お手数ですが再度確認お願いします～🙇‍♂️

o-musubi committed:
軽微な名前の修正

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,29 @@
+  file[:splited_columns] = file[:list].each_slice(file[:column_difference]).to_a
+  file[:specific_case] = MAXIMUM_COLUMNS - file[:splited_columns].size
+  file
+end
```
やっぱりメソッド設計が恣意的なものになっていて、 `find_conditions` が何をするものなのか、あまりしっくりこないです。後続の処理のための必要な変数の準備をただエイヤとひとくくりにしただけに見えてしまいますね。
前回のコメントでも書きましたが、一旦実装してみて必要な処理は見えてきたと思うので、トップダウンな考え方で整理したほうがよいかと思います。

例えば、カレーの作り方をプログラムとして捉えると

```
- カレーを作る
  - 材料の下ごしらえ
    - 玉ねぎを切る
    - にんじんを切る
    - じゃがいもを切る
  - 材料に火を通す
    - 肉、野菜を炒める
    - 水を入れて煮込む
  - 味つけをする
    - ルーをいれる
```

のような工程になると思います。
この各項目の粒度をそろえてメソッド化すると見通しがよいです

```ruby
def make_curry
  prepare_ingredients
  cook_ingredients
  season
end

def prepare_ingredients
  chop_onions
  # ...
```

場当たり的にメソッドを作っていると

```ruby
def make_curry
  chop_onions
  chop_carrots_and_potatoes  
  cook_ingredients
end
```

のように粒度がまちまちになって、読みづらいコードになりがちです。
最初から綺麗にトップダウンに設計することは僕でもなかなか難しいですが、だいたい処理が見えてきたらトップダウンに一度整理してみると、コードの見通しを立てやすいです。

o-musubi replied:
メソッドの粒度を意識して、以下の形で処理を行えるようメソッドの設計を考え直しました。

1.  ファイルを取得する
2.  ファイルを出力する状態に整える
    1. 画面に表示する配列の左右の差を計算する(全体の要素数÷出力したい列数、切り上げ)
    2. ファイルを分割する(各配列の同じ添え字の要素同士の差が、画面に表示する列の差と一致するように)
     3. 画面に表示する順番に並べ替えた配列を作成する
     4. 2-3で作成した配列を、画面に表示する幅に分割する
    ※ただし、2-2で分割した配列の数が最大出力幅を下回る場合は処理が通常と異なるので注意
3.  出力する
    1. 表示幅を計算する
    2. 2-2で作成した配列を順に画面に表示

yoshitsugu reviewed (commented):
コメントしました！

o-musubi committed:
メソッド設計の見直し

o-musubi committed:
軽微な変更をmerge

o-musubi commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,43 @@
+  end
+end
+
+def format_files
```
`format_file`の部分について、自分では改善方法が思いつかなかったので教えてください。

現在、"ファイルを整形する"という処理は、ファイルを受け取ってから出力する状態の配列を作成するまででひとくくりだと考えています。
しかし、提出した状態だと、メソッドとしてはコードが長いと思います。

この場合、考えられる対策としては
1．メソッドとして抜き出す
2. コードを単純に短くする
のどちらかだと思います。

並べ替えた配列を再分割する部分を、メソッドとして抜き出してみると、戻り値を使うためにfiles[:キーワード]が多くなってまい、読みづらい気がしました。また、ご指摘いただいた処理の粒度？がそろわなくなってしまう気がします。(あくまでファイルを整形する途中の処理なので)

ならばコードを短くすればいいのかと、三項演算子など試してみたのですが、その後の処理が長いためかえって読みづらいかな？と思いました。他に短くする方法はどういったものがありますか？

そもそも、設計がまだおかしいでしょうか…？
アドバイス頂けると嬉しいです～🙇‍♂️

yoshitsugu replied:
いくつかやり方はありますが、今回の場合、アルゴリズムを改善してみるのがよいかと思います。
処理を見てみると

1. 3列だと何行になるか？を計算(X行になるとする)
2. X個ごとにグルーピング
3. 3つごとにグルーピングしたときに並び順が正しくなるように列び替え
4. 3つごとにグルーピング

となっているようですね。できるだけこの処理を活かすとすると、3, 4の手順はまとめられそうです。

```ruby
file_nemes_to_display = []
chunk = [] # 3つごとにわけるための一時的な配列
splited_columns.cycle(column_difference).with_index do |splited_column, i| 
  chunk << splited_column.shift
  if i % MAXIMUM_COLUMNS == 2
    (file_names_to_display に chunkの内容を追加して、chunkをリセット)
  end
end
```

また、細かい見やすさでいうと、たとえば

```ruby
def column_size(file_names)
  file_names.size.ceildiv(MAXIMUM_COLUMNS)
end
```

のような計算するメソッドを作っておくと少し見通しがよくなります。
ただ、あまり細かくメソッドを作りすぎると、結局何をしているメソッドなのか、定義を確認するために視線が上下することを強制するだけの分け方になってしまうので、「ここは抽象的にいうと何をしているのか？」「目的は何か？」という観点でまとめるとよいです。

o-musubi replied:
アルゴリズム、指示いただいたように修正しました。
特例(3列表示の時のファイル数4、4列表示の時のファイル数5,6,9等、最大出力幅を下回って出力する場合)についての分岐がなくなっているのは、その出力に関してはメソッドを分けた方がいいということでしょうか？
今同じメソッド内で処理していて、確かに条件式が長くなってしまっているのですが、インデックスとリンクしているので切り離しがうまくいきませんでした…。
再度アドバイスいただけませんでしょうか。

o-musubi commented:
レビューをもとに設計を見直しました。
お手数ですがご確認よろしくお願いします！

o-musubi committed:
print_filesメソッドととmainメソッドを結合

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,39 @@
+  files
+end
+
+def find_files
+  {
+    list: Dir.glob('*')
+  }
+end
```
ここではHashにいれる必要はないかなと思います。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+def main
+  files = format_files
+  width = files[:list].map { |a| a.to_s.bytesize }.max
+  files[:displayed_files].each do |displayed_files|
+    displayed_files.each do |displayed_file|
+      printf("%-#{width}s\t", displayed_file)
+    end
+    puts ' '
+  end
+end
```
`format_files` は「渡されたファイル名をフォーマットするメソッド」として単機能にしたほうがよいと思いますので、

```ruby
def main
  file_names = find_files
  formatted_file_names = format_file_names(file_names)
```

などとできるようにしたほうがよいでしょう。
(ここでは `format_files` を `format_file_names` という名前に変更しました。)
こうすると `format_files` 側は渡されたものに対して出力するだけなので、「ファイル名の取得には `find_files` というメソッドを呼び出さなければならない」という知識も知らずに済みます。
オブジェクト指向のプラクティスで詳しくは学ぶとよいですが、こうやってメソッドやクラスの独立性をあげていくと、保守しやすいコードになりやすいです。

yoshitsugu reviewed (commented):
コメントしました！

o-musubi committed:
全体の設計を修正

mainメソッドとdisplay_filesを分離
format_file_namesの計算部分をメソッドに抜き出し
format_file_namesの繰り返し部分の処理をを変更
不要なHashを削除

o-musubi committed:
format_file_namesとswap_file_namesを分離

o-musubi commented:
基本的に1つのメソッドは1つのことしかしない、という方針で設計しなおしました。(それに伴いメソッドの呼び出し順が変わったので配置も変えました)
一つ一つのメソッドが何をしているかや、メソッド同士の関係性は以前より明示的になったと思うのですが、コードの長さ自体はかなり長くなってしまいました。問題ないでしょうか…？

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,54 @@
+  end
+end
+
+MAXIMUM_COLUMNS = 3
```
細かいですが、定数宣言はファイルの上のほうにまとめてあるか、使用される箇所のすぐ上にあると、参照しやすいですね。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,54 @@
+  file_names_to_display = swap_file_names(difference, columns)
+end
+
+def column_difference(file_names)
```
differenceというと 「差」や「違い」のようなイメージなので、単に「column_size」とかでもいいのかなと思いました。
また、ここは列の数が決まっている状態で、行の数を求める処理かなと思うので、 `row_size` かもと思いました。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,54 @@
+def format_file_names(found_file_names)
+  file_names = found_file_names
+  difference = column_difference(file_names)
+  columns = splited_columns(file_names, difference)
```
`splited`は  `splitted` のスペルミスですかね〜
引数に対して何か処理をして返すようなメソッドは動詞を使うことも多いです。
`split_into_columns` とかでもよいかと思います。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,54 @@
+  file_names_to_display = swap_file_names(difference, columns)
+end
+
+def column_difference(file_names)
+  file_names.size.ceildiv(MAXIMUM_COLUMNS)
+end
+
+def splited_columns(file_names, difference)
+  file_names.each_slice(difference).to_a
+end
```
このあたりはわざわざメソッドとしてわけるまでもないかもしれません。

```ruby
row_size = file_names.size.ceildiv(MAXIMUM_COLUMNS)
splitted_rows = file_names.each_slice(row_size).to_a
```

と変数名を適切につければ、何のためにやっている処理なのかもある程度わかりますしね。

yoshitsugu reviewed (commented):
だいぶ設計としてはよくなりました :+1: 
細かいところ中心にコメントしました！

o-musubi committed:
軽微な修正(定数の位置、メソッド名、不要なメソッドと変数の削除)

o-musubi committed:
変数名を修正

o-musubi commented:
修正しました！再度確認お願いします🙇‍♂️

※2回目のpushのコメント、変数名ではなくメソッド名です。すみません！

yoshitsugu reviewed (approved):
OKです！
